### ISSUE TITLE
Tuya -> Tasmota/ESPHome -> HomeAssistant and beyond

### ISSUE THREAD

**Issue by deftdawg** [gknepper/Smart-Plug---CE---HPPA11SW#1]
Picked up a pack of these last week, I'm impressed they are 15A and CSA marked for $20 CAD...

Have you had a chance to open one yet?  Most of these devices seem to be tuya / Esp based controlled, I'm curious if this is as well.  Would be good to get some picks of the board chips, pins, and general layout to see if it's already been converted by someone else under a different brand name.

I did grab the APK and unzip it: [com.homedepot.sp1_131_apkplz.net.apk](https://apkplz.net/app/com.homedepot.sp1)

App is written in Javascript, there are a ton of different devices in /assets and build.zip seems to contain the whole app, though it's been run through webpack. 

Pasted com.homedepot.sp1_131_apkplz.net.apk/assets/static/js/wifi-devices.2dbbbda4.chunk.js into https://lelinhtinh.github.io/de4js/ and there are references to port 6666 and 6667 on 255.255.255.255 ... though the code is still impossible to read. 

Anyway, I'd love to get these running on ESPHome or Tasmota so I can control them directly via [HomeAssistant](https://www.home-assistant.io/).

**Comment by gknepper** [comment]
Hi @deftdawg , Actually I opened one already. The images are in images folder. It runs on a RTL8710BX chipset, not on a ESP82XX. So the device it's **not** compatible with Tasmota which runs only on ESP8266 or ESP8285 chipsets. I sniffed the packages that this device receives and it's all MQTT packages. (with SSL). so it's a remote MQTT subscription that controls the relay.

The board has 4 pins marked as GND/+3V3/RX/TX where I want to try to get a firmware dump. ( some soldering job is needed and I don't have a soldering iron available ) but it's an alternative to understand the working of this device.

**Comment by deftdawg** [comment]
Ah yes, I started looking into this and it seems starting with Teckin sp22 plugs many manufactures have started switching to Realtek Arm chips.

There are some indications that these can be programmed like Ardunios
- [Ameba Writer, easy programming interface for RTL8710 on Arduino IDE](https://www.youtube.com/watch?v=Qc4NaQRBFnM) & [Ameba Writer](https://www.tindie.com/products/microwavemont/ameba-rtl8710-programmer-for-arduino-ide/)
- [platformio-realtek-rtl8710b - PlatformIO development system platform for realtek microcontrollers.](https://github.com/8devices/platformio-realtek-rtl8710b)

Here are some other links I found:
- [tuya_iot_sdk_for_RTL8710BN - OTA](https://github.com/TuyaInc/tysdk_for_rtl8710bn/blob/master/wifisdk_for_rtk_871x/component/soc/realtek/8711b/misc/rtl8710b_ota.c)
- [Tuya WR2 Developer Datasheet](https://developer.tuya.com/en/docs/iot/device-development/module/wifi-module/wr-series-module/wifiwr2module?id=K9605tko0juc3)
- [Hackaday: RTL8710 easy programming by Arduino IDE](https://hackaday.io/project/19163-rtl8710-easy-programming-by-arduino-ide)
- [Some SDK by pvvx](https://github.com/pvvx/RTL0B_SDK)

BTW, Is there a non-destructive way to get it out of the casing?   It looks like there's a seam along the back, is it clipped only or glued as well?

**Comment by gknepper** [comment]
Probably, I manage to open with a minimal damage on the case. I don't have but this tool can probably solve the issue: 

https://www.amazon.ca/iFixit-Jimmy-Electronics-Opening-Tool/dp/B00NCFIVH4

It's only 4 clips but the way they did make it hard. I'll include some images later on.

**Comment by deftdawg** [comment]
> It's only 4 clips but the way they did make it hard. I'll include some images later on.

Yeah some pics of the inside of the shell (or what's left of it 🤣) would be great.  Maybe can get it open with a putty knife or something...

**Comment by deftdawg** [comment]
Not sure if it's useful, but seems to be a Russian firmware building site for RTL8710n here...   https://wifi-iot.com/p/rtl87xx/

Login is demo/demo

Most of the features (red) appear to require payment.

![image](https://user-images.githubusercontent.com/4991612/103307661-9ec1cb00-49de-11eb-82ad-b200c6144865.png)

**Comment by gknepper** [comment]
I would say worth a try, but without a better understanding on the firmware dump/written process it's quite a leap. I still don't know if an ameba writer is needed or only a serial connection can get the job done.

**Comment by deftdawg** [comment]
I'm also unsure...  There is a discussion here: https://github.com/arendst/Tasmota/discussions/9262

Someone drops a link [Ameba ARDUINO: Getting Started with RTL8710](https://www.amebaiot.com/en/ameba-arduino-getting-started-rtl8710/), but appears to be an official SDK board or something.

**Comment by gknepper** [comment]
@deftdawg I upload some photos of the case opened and my concerns about the way they create some sort of groove in the plastic.

**Comment by funroompc** [comment]
Best of luck guys. I’m following this :)

**Comment by gmcmicken** [comment]
> I sniffed the packages that this device receives and it's all MQTT packages. (with SSL). so it's a remote MQTT subscription that controls the relay.

@gknepper How far did you get in capturing the mqtt packets, could you read the contents, is it a known certificate?

**Comment by gknepper** [comment]
@gmcmicken  We are in the same page pretty much. It's a proprietary certificate for sure. I was trying to dump the  RTL8710BX image and with some luck open and check the contents to find the cert. Some soldering is needed and I haven't done so far.

**Comment by gmcmicken** [comment]
@gknepper I found the keypair, it was in the apk in a keystore file (.bks) and I guessed the password. 👍 

[leedarson-keypair.zip](https://github.com/gknepper/Smart-Plug---CE---HPPA11SW/files/5854485/leedarson-keypair.zip)

**Comment by gknepper** [comment]
@gmcmicken Sweet!

**Comment by gmcmicken** [comment]
I tried packet capturing tonight, but my wifi card doesn't support monitor mode. I then tried to capture on my android device while using the app to configure the smart plug but the app failed to start when the capture was running (an android vpn issue I guess?) I'll try again though.

**Comment by gknepper** [comment]
I tried using rpcapd inside my router but I wasn't able to open the packages. I think I'm doing something wrong on wireshark on TLS settings. For now my only findings was the MQTT gateway: azuremqtt.arnoo.com which leads us to the probable manufacturer of this Arnoo app. https://www.arnoo.com/

**Comment by gmcmicken** [comment]
It's possible the keypair I found is only for the intial setup during the stage where the switch is broadcasting an unsecured wifi and the app connects directly to it. It's also possible the keypair is for some other device supported by that app. I think we should be able to see the fingerprint (?) in the tls handshake though to be sure.

**Comment by GSzabados** [comment]
I know this is not the same device, but some hint where to look for some ideas...

https://www.reddit.com/r/homeautomation/comments/j5kb2o/hacking_a_7_zigbee_mqtt_hub_that_comes_with_two/

**Comment by bluesky-ca** [comment]
FCC info for the company making the reference design: [https://fccid.io/2AB2Q](url), and this specific plug [https://fccid.io/2AB2QHPPA11SWB](url) has anybody been able to find more info on local control ?

**Comment by sree-86** [comment]
Any update on this?

**Comment by eastpole** [comment]
At long last I think some cool people got traction on this problem. 
[    https://github.com/libretiny-eu/libretiny](    https://github.com/libretiny-eu/libretiny)
    
  This project is enabling the creation of firmware that can be run on non-ESP32 IoT platforms including the relevant one here, the RTL8710BX. I really did not want to get tangled up with Ameba if I could help it -- this looks more likely to work for me.

**Comment by bluesky-ca** [comment]
I wonder if anybody has actually managed to flush a plug using that code - the project does look interesting.

For a while now I have been using the Sonoff S31 plugs with tasmota, but there is a big price difference between the HomeDepot and the Sonoff ( including the lite version without power monitoring ).  
I did some hacking at the plug while back using ssl proxy but was not able to obtain local control - was using an older rooted phone, but have seen videos using the Android emulator making the process a bit simpler.  

I believe HD still sells those plugs at a good price, so it would be great to have a low cost, easy to purchase option with local control ( checked the HD page and they still do sell them ).

**Comment by ESurge** [comment]
TLDR: I have this switch working with ESPHome. You will need to...
- solder some wires to TX0, RX0, and GND
- manually compile/download the OTA1/OTA2 files using ESPHome
- use [rtltool.py](https://github.com/pvvx/RTL0B_SDK/blob/master/mkb/rtltool.py) for flashing OTA1/OTA2 (instructions below)
- use the YAML code (below) for ESPHome or use the following pin assignments
  - Button: PA22
  - Relay: PA00
  - Green LED: PA05
  - Red LED: PA14
  - LibreTiny Module: BW12
-------------------------------------------------------------------
I bought a three pack of switches on sale at Home Depot. At the time, there wasn't any information for the RTL8710 so I converted two switches with ESP8266 chips for an _almost_ drop-in replacement. Flashed Tasmota on them and went on my way.

I had one switch remaining and until recently was just put aside. I was about to convert it, but I found out that the LibreTiny/ESPHome project exists which has made progress with RTL8710 chips.

LibreTiny mentions how to get a module setup in ESPHome, and then to use the **ltchiptool** for flashing. The ltchiptool helped me figure out that these SP1 boards have 2MB of flash. Using a generic rtl8710bx module configuration and no GPIO settings, I compiled a firmware. Sadly, my attempts to flash a UF2 firmware file caused ltchiptool to crash or not write to the chip.

I found another tool called **rtltool.py**. This tool plus the ability to manually download the OTA1/OTA2 files allowed me to flash the board. After this success, I manually figured out which pin assignment was associated to the components which I had previously reversed engineered from converting my other boards over to ESP8266.

Notes:
- I'm new to ESPHome and figured this out in the last day and a half, so this might not be the best but it's working for me.
- I can't use ESPHome to flash updates. I'm not sure how to get this working if at all.
- `rtltool.py` requires **Python 2.7** to work
- When modifying the commands listed below, make sure to change the USB to Serial device port, and also ensure you get the flash addresses correct.
- I noticed that the SP1 doesn't always connect to wifi. I'm not sure if this is a consequence of using the BW12 LibreTiny module or maybe my wireless connection is spotty.

To flash the RTL8710BX chip:
**Manually download the OTA1/OTA2 bin files from ESPHome before proceeding.**
  - I used a power bar/strip with a on/off switch
  - With the power bar off, plug in the SP1
    - The top cover was off, exposing the board
    - This is dangerous so proceed with caution!
    - Edit: You can do this more safely instead of connecting to power, use an external 3.3V source and connect it to the 3.3V pin on the back of the board.
  - On the USB to Serial device:
    - Connect RX0 to TX 
    - Connect GND to GND
    - Connect TX0 to GND
      - This is required to bring the RTL8710BX into Download Mode
  - Power on the SP1
  - Change the TX0 pin over to RX on the USB to Serial device
  - In a command terminal:
    - Ensure the OTA1/OTA2 bin files are in the current working directory otherwise you'll need to specify the exact location
    - Modify and run the following command to write to OTA1:
      - `python rtltool.py -p <USB_to_Serial_Port> wf 0xb000 name_of_device_0x00B000.bin`
    - Modify and run the following command to write to OTA2:
      - `python rtltool.py -p <USB_to_Serial_Port> wf 0x80000 name_of_device_0x080000.bin`
  - After flashing, power off the SP1, and then power on again
  - You should now have ESPHome flashed to the SP1

YAML:
```
esphome:
  name: name_of_device
  friendly_name: "Name of Device"

rtl87xx:
  board: bw12

# Enable Home Assistant API
api:
  encryption:
    key: "keyhere"

ota:
  password: "ota_pass"

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

binary_sensor:
  - platform: gpio
    pin:
      number: PA22
      mode:
        input: true
        pullup: false
      inverted: true
    name: "Button"
    on_press:
      - then:
        - switch.toggle: relay
    disabled_by_default: true

switch:
  - platform: gpio
    name: "Relay"
    pin: PA00
    id: relay
    on_turn_on:
      - then:
        - output.turn_on: green_led
    on_turn_off:
      - then:
        - output.turn_off: green_led

status_led:
  pin:
    number: PA14
    inverted: false

output:
  id: green_led
  platform: gpio
  pin: PA05
```

Some images:
![IMG_20240212_200829](https://github.com/gknepper/Smart-Plug---CE---HPPA11SW/assets/982575/bc8b15d7-ee5f-421d-b849-bb0ef2729793)
![IMG_20240212_194920](https://github.com/gknepper/Smart-Plug---CE---HPPA11SW/assets/982575/6f180539-3ef7-4b61-8b44-a2b3333c3839)

**Comment by bluesky-ca** [comment]
Thanks for the info.
Would be interesting if one could flash the plug over the air without having to crack it open.   Even without that being able to flash it with ESPHome is great progress - would love to see Tasmota but I don't believe they are doing devices outside of ESP chips.   I guess it is the time to cut them open :-)

**Comment by ESurge** [comment]
Ideally that would be best (flashing without having to tear apart the case). Perhaps when more research into how the original firmware operates is done but until then, I'll take what I can get.

I found a way to connect 3.3V and GND without having to take the screws out. The empty capacitor spot by the square cutout has GND (white striped/shaded section), and 3.3V (blank section) pins.

Image attached.
![IMG_20240215_205812](https://github.com/gknepper/Smart-Plug---CE---HPPA11SW/assets/982575/215c279e-40ef-4a8e-9f95-03f85fff6b67)

**Comment by ESurge** [comment]
Grabbed a few more 3pk boxes from Home Depot since they were on clearance.

When flashing one of these new devices, the button (PA22) didn't work with the ESPHome YAML I posted above. To fix this, I had to enable the internal pullup resistor.

Regarding the button (PA22) pullup resistor, I've seen a mixed result; detailed below.

Mostly I've been flashing with the pullup set to false. This is because when I was first experimenting to get the correct pins and settings for the YAML, I noticed the button wouldn't work unless the pullup was set to false.

After flashing a few of these, I ran into the device first mentioned above where the pullup needed to be set to true and didn't work otherwise.

I also tested some devices by leaving the pullup set to true. This worked on some devices but not others. Strange.

**Comment by ESurge** [comment]
A few months later, I am more familiar with ESPHome, Home Assistant, and these RTL8710 chips.

LibreTiny has made some updates. I'm currently using libretiny v1.5.1, and ltchiptool v4.11.1. ltchiptool can now flash these chips using the UF2 file format. In addition, after re-flashing with the UF2 file, wireless updates are now working over ESPHome Dashboard.

Unfortunately, you'll still need to open these smart plugs up and manually flash the first time. Also, if you've flashed with the `rtltool.py` script I mentioned above, you'll need to reflash since this method didn't let wireless updating work.

Some notes:
- Logger works now so adding `logger:` to the YAML won't break the device
- Wireless updating works over ESPHome
- I used Linux (a Ubuntu 20.04 VM) and ltchiptool via command line to flash
  - Command used: `ltchiptool flash write filename.uf2`
  - UF2 file is generated by ESPHome using the YAML template I posted [here](https://github.com/gknepper/Smart-Plug---CE---HPPA11SW/issues/1#issuecomment-1940440953)
- Tested on two devices, and so far so good

Issues (New and Remaining):
- I'm not sure if it's because of my USB to UART device or something else, but reading the "chip info" takes a very long time.
  - If I read the chip info specifically, on Linux or on Windows, it seems to take hours to complete.
  - On Windows, any command ran with ltchiptool initiates the chip info query once `Transmission successful` with the chip is established. This means any commands could takes hours if it doesn't end up failing (usually).
  - On Linux, this doesn't seem to be the case, and I can flash the UF2 file in about 2 minutes.
- Sometimes Wi-Fi doesn't get picked up. Power cycling usually fixes this. However, this was seen before and is probably due to being a low powered device.

**Comment by Deep-Six** [comment]
@ESurge  awesome timing, I just found a few of these in an old box. Do you think it's likely that it'll be possible to flash them without cracking them open soon? I remember just how difficult they were to get into before, so wondering if I can just wait and there'll be a OTA ability soon?
Thanks!

**Comment by bluesky-ca** [comment]
Its always nice to see progress on projects like this - thanks for all the updates.   
Has anybody been able to extract the original firmware and possibly run binwalk on it ?

**Comment by ESurge** [comment]
Not sure about wirelessly updating the stock firmware to ESPHome. I think I tried the Tuya CloudCutter tool in Feb 2024, but I can't remember. If it succeeded, I would've posted about it.

I've attached the stock firmware that I managed to dump. I haven't taken a look but really don't know where I would start.

[Consumer Electric SP1 Smart Plug Stock Firmware Dump.zip](https://github.com/user-attachments/files/16384568/Consumer.Electric.SP1.Smart.Plug.Stock.Firmware.Dump.zip)

**Comment by bluesky-ca** [comment]
Looks like that dump is not the original firmware - there are esphome strings (esphome v2023.12.9) in it and is dated Feb 21 2024, 17:17:32 ;-)

**Comment by ESurge** [comment]
Yep, I deleted that post. I swear I checked the file before hand, but still ended up uploading a file with some credentials on it :(

**Comment by ESurge** [comment]
Not sure about wirelessly updating these from stock firmware to ESPHome. This would be a nice feature to have.

I can't recall if I tried Tuya CloudCutter tool that is posted on LibreTiny's Getting Started page. However, if it worked, I would've posted about it.

Anyways, apparently I lost my SP1 Stock dump, and I didn't have another stock device available. Therefore, I went and got another one at Home Depot today and dumped it.

[Commercial Electric SP1 Smart Plug Stock Firmware Dump.zip](https://github.com/user-attachments/files/16387433/Consumer.Electric.SP1.Smart.Plug.Stock.Firmware.Dump.zip)

**Comment by Deep-Six** [comment]
I tried cloud cutter last night, but there's no profile for any CE devices.
